### ISSUE TITLE
fn.branching.conditional with mesh or swarm variables....

### ISSUE THREAD

**Issue by rbeucher** [underworldcode/underworld2#207]
I came across some unexpected behaviour while trying to push the logic of mesh and swarm variables being underworld functions.

Note that the example below is completely arbitrary.... It is just intended for testing...

First, I naively thought that I could build a condition such as:

``` python
conditions = [( materialIndexField == 0, 2), (True, materialIndexField)]
materialIndexField.data[:] = fn.branching.conditional( conditions ).evaluate(swarm)
```
materialIndexField being a swarmvariable attached to swarm.

I am using those kind of things in my models and it took me some time to realize that the resultant of the last clause was systematically returned...it is potentially quite dangerous...
The variable "materialIndexField" is evaluated when it is in the "resultant" but not when it is in the "condition"...

So based on my understanding of the code, each condition is converted to an underworld function using function.convert. The function works correctly when dealing with normal underworld functions but not for mesh or swarm variables.
In short, if you do fn.Function.convert( materialIndexField == 0) this will always evaluate as a fn.misc.constant initialized to False (variable == int is always False)...

I think that is an issue... There are some workarounds for that but I quite like the idea of building conditional function like that... 
I use them for building friction weakening or cohesion weakening functions based on the value of accumulated plasticstrain...

I tried:

``` python
conditions = [( materialIndexField.evaluate(fn.input()) == 0, 2), (True, materialIndexField)]
materialIndexField.data[:] = fn.branching.conditional( conditions ).evaluate(swarm)
```
but it does not help....(was a bit of a leap of faith :-))

Romain

**Comment by rbeucher** [comment]
Here is a complete minimal example:

```python
import underworld as uw
import glucifer
import underworld.function as fn

mesh = uw.mesh.FeMesh_Cartesian(elementType = ("Q1/dQ0"),
                                elementRes  = (20, 10),
                                minCoord    = (0., 0.),
                                maxCoord    = (2., 1.),
                                periodic    = (False, False))

swarm  = uw.swarm.Swarm( mesh = mesh, particleEscape=True)
swarmLayout = uw.swarm.layouts.GlobalSpaceFillerLayout( swarm = swarm, particlesPerCell=50 )
swarm.populate_using_layout( layout = swarmLayout )

materialIndexField = swarm.add_variable( dataType="int", count=1 )

coord = fn.input()
conditions = [(coord[1] > 0.5, 0), (True, 1)]
materialIndexField.data[:] = fn.branching.conditional( conditions ).evaluate(swarm)

conditions = [( materialIndexField == 0, 2), (True, materialIndexField)]
materialIndexField.data[:] = fn.branching.conditional( conditions ).evaluate(swarm)

figMat = glucifer.Figure((1200, 600))
figMat.append(glucifer.objects.Points(swarm, materialIndexField, fn_size=3.0))
figMat.show()
```

**Comment by jmansour** [comment]
As you point out, we don't overload the == operator for functions, hence you are using a plain python == operation, which will of course evaluate to False.  I'm not what the solution is here.. perhaps a dummy == overload which flags the error, though this would prevent usage in the python sense. 

I originally included an == overload, but the issue is that currently the logical function operations only really do floating point comparison, and doing == of floating point objects is often dangerous.  I would like to expand functions to handle integer comparison natively, but this requires some work.  Note that this will work:

```
conditions = [( materialIndexField < 0.1, 2), (True, materialIndexField)]
materialIndexField.data[:] = fn.branching.conditional( conditions ).evaluate(swarm)

figMat = glucifer.Figure((1200, 600))
figMat.append(glucifer.objects.Points(swarm, materialIndexField, fn_size=3.0))
figMat.show()
```

Or you might like to define a helper function:
```
def isequal( func_guy, value, delta=0.01):
    return (func_guy < value+delta) & (func_guy > value-delta)

conditions = [(isequal(materialIndexField,0),2), (True, materialIndexField)]
materialIndexField.data[:] = fn.branching.conditional( conditions ).evaluate(swarm)

figMat = glucifer.Figure((1200, 600))
figMat.append(glucifer.objects.Points(swarm, materialIndexField, fn_size=3.0))
figMat.show()
```

There is a performance cost in the extra evaluations, but I'm guessing this won't be an issue.
